File: content/LLD/Reference-Code/20. Facade Pattern/good_example.py
class UserService:
    def login(self, username: str, password: str) -> dict:
        print(f"[UserService] Logging in: {username}")
        return {"user_id": "U123", "name": username}

    def get_profile(self, user_id: str) -> dict:
        print(f"[UserService] Getting profile for: {user_id}")
        return {"user_id": user_id, "name": "Rahul", "address": "Mumbai"}


class OrderService:
    def get_orders(self, user_id: str) -> list:
        print(f"[OrderService] Getting orders for: {user_id}")
        return [
            {"order_id": "ORD-1", "total": 1500},
            {"order_id": "ORD-2", "total": 3000},
        ]


class ApiGateway:
    def __init__(self):
        self.__user_service = UserService()
        self.__order_service = OrderService()

    def login_user(self, username, password):
        self.__user_service.login(username, password)

    def get_user_profile(self, user_id):
        self.__user_service.get_profile(user_id)

    def get_order_details(self, user_id):
        self.__order_service.get_orders(user_id)

    def get_all_details(self, user_id, username, password):
        self.__user_service.login(username, password)
        self.__user_service.get_profile(user_id)
        print(self.__order_service.get_orders(user_id))


api_gateway = ApiGateway()
api_gateway.get_all_details("d23e32", "test", "123")

File: content/LLD/Reference-Code/20. Facade Pattern/GoodExample.java
import java.util.List;
import java.util.Map;

public class GoodExample {
    static class UserService {
        public Map<String, String> login(String username, String password) {
            System.out.println("[UserService] Logging in: " + username);
            return Map.of("user_id", "U123", "name", username);
        }

        public Map<String, String> getProfile(String userId) {
            System.out.println("[UserService] Getting profile for: " + userId);
            return Map.of("user_id", userId, "name", "Rahul", "address", "Mumbai");
        }
    }

    static class OrderService {
        public List<Map<String, Object>> getOrders(String userId) {
            System.out.println("[OrderService] Getting orders for: " + userId);
            return List.of(
                Map.of("order_id", "ORD-1", "total", 1500),
                Map.of("order_id", "ORD-2", "total", 3000)
            );
        }
    }

    static class ApiGateway {
        private final UserService userService = new UserService();
        private final OrderService orderService = new OrderService();

        public void loginUser(String username, String password) {
            this.userService.login(username, password);
        }

        public void getUserProfile(String userId) {
            this.userService.getProfile(userId);
        }

        public void getOrderDetails(String userId) {
            this.orderService.getOrders(userId);
        }

        public void getAllDetails(String userId, String username, String password) {
            this.userService.login(username, password);
            this.userService.getProfile(userId);
            System.out.println(this.orderService.getOrders(userId));
        }
    }

    public static void main(String[] args) {
        ApiGateway apiGateway = new ApiGateway();
        apiGateway.getAllDetails("d23e32", "test", "123");
    }
}
